cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java

```diff
@@ -60,21 +60,25 @@
   private final DiskFailureDetector _diskFailureDetector;
   private final ScheduledExecutorService _detectorScheduler;
   private final long _anomalyDetectionIntervalMs;
-  private final LinkedBlockingDeque<Anomaly> _anomalies;
+  private final PriorityBlockingQueue<KafkaAnomaly> _anomalies;
   private volatile boolean _shutdown;
   private final LoadMonitor _loadMonitor;
   private final AnomalyDetectorState _anomalyDetectorState;
   private final List<String> _selfHealingGoals;
   private final ExecutorService _anomalyLoggerExecutor;
-  private volatile Anomaly _anomalyInProgress;
+  private volatile KafkaAnomaly _anomalyInProgress;
   private volatile long _numCheckedWithDelay;
   private final Object _shutdownLock;
 
   public AnomalyDetector(LoadMonitor loadMonitor,
                          KafkaCruiseControl kafkaCruiseControl,
                          Time time,
                          MetricRegistry dropwizardMetricRegistry) {
-    _anomalies = new LinkedBlockingDeque<>();
+    // For anomalies of different types, prioritize handling anomaly of higher priority;
+    // otherwise, handle anomaly in order of detected time.
```

Comment:
I believe this _best-effort_ prioritization (because except for the broker failure detection, all other anomalies are detected periodically) should be avoided as 
1. the detection logic for each anomaly (should) already handle when to generate an anomaly in a best effort fashion -- e.g. please see (1) `GoalViolationDetector#shouldSkipGoalViolationDetection` (2) `DiskFailureDetector#shouldSkipDiskFailureDetection` and (3) `MetricAnomalyDetector#shouldSkipMetricAnomalyDetection`
2. it may cause starvation of certain anomaly types,
3. deprioritized anomalies will become irrelevant as their priority is dropped in the queue -- e.g. when would it be helpful to handle a `goal_violation` after handling a `broker_failure` that occurred after it? a broker failure self-healing will start a long-running execution and by the time the goal violation is picked from the anomaly queue, goal violation anomaly will be disposed due to the ongoing execution.
4. this prioritization makes the debugging harder.

Reply:
The main reason I make this change is for scenario like the following.
1. A goal violation is detected at time 100
2. A slow broker is detected at time 101
3. AnomalyHandler picks up the next anomaly to work on

In this scenario, I want the slow broker to be picked up first because if CC first works on goal violation and trigger a long-time rebalance operation, it will not help slow broker issue and the existence of slow broker may cause backlog in the producer and cascade issue.
Therefore, it is better to first demote the slow broker(which will be fast), mark it as recently demoted broker, then do rebalance(thus no new leader will be moved to this slow broker).

For the concerns above, here are some of my thoughts.
1. This is preempting mechanism in anomaly detection, the priority queue gives us a similar preempting mechanism in anomaly handling. 

2. I feel this is ok, because for one thing, the high priority anomaly means more serious ones and deserves being addressed first; for another thing,  when CC is doing `remove_broker`(for broker failure or slow broker) or `fix_offline_replicas` (for disk failure), it actually runs customized rebalance, i.e. it still use the same set of goals thus helps to remedy goal violation(the starved type).

3. in that example, the order does not matter. but I think in the scenario I described above, the order change matters. In both scenario, changing to priority queue does not make the case worse.

4. I feel that is acceptable, as long as we can get the history of anomaly handling via [log](https://github.com/linkedin/cruise-control/blob/e73256a79d6ef9c3c1063452dd5e9eacd02e7d1c/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java#L419) and [operation log](https://github.com/linkedin/cruise-control/blob/db6281c2c458ded9dddad72606048bcc01f135ae/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java#L696).

Reply:
If slow broker handling is intended have a higher priority, then would scheduling its periodic check before goal violation checks help? -- i.e. in `AnomalyDetector#startDetection`, for `scheduleAtFixedRate` call can we set a _sufficiently_ earlier initial delay to set higher priority anomaly detectors?
As long as the `slow-broker-anomaly-generation-time` is less than the `difference between the other lower priority anomaly detector's start time` + `other-detector's-anomaly-generation` time, it will have higher priority. Note that we use the same frequency in (1) slow broker, (2) disk failure, and (3) goal violation detection; hence, the period of anomaly detection will stay fixed. Moreover, if the intention is to capture slow brokers with higher frequency, can't we run its detector more frequently, as well.

Reply:
Delay the detection start time only helps to some extent. 
The priority does not maintain after a self-healing operation because self-healing can take arbitrary time. It could be
 `slow broker detection skipped` -> `self-healing finished` -> `goal violation triggered`

Increase detection frequency sounds like a good idea. I will add configs for each anomaly type.